### PR TITLE
[ncl] add ApplicationScreen to ncl based on expo-application

### PR DESCRIPTION
**Pull request by vivianzzhu91**
# Why

Created the `expo-application` module and want to add the `ApplicationScreen` to NCL to test those functions.

# How

Add a new `ApplicationScreen`  to NCL and test all functions and constants of the `expo-application` module.

# Test Plan
 
Tested both on Android and iOS simulators and test each constant and function outputs.
WechatIMG2
WechatIMG1
Screen Shot 2019-07-19 at 4 00 18 PM
Screen Shot 2019-07-19 at 4 05 17 PM

### CODE REVIEW

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    }
+  }
+}
+
+export default class DeviceScreen extends React.PureComponent<props, mState> {
+  static navigationOptions = {
+    title: 'Device',
+  };
+
+  constructor(props: Readonly<props>) {
```
Comment (ide) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]: remove the construct and just use a property initializer for `state`

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+interface mState {
+  installReferrer?: string;
+  idForVendor?: string;
+  firstInstallDate?: string | undefined;
+  lastUpdateDate?: string | undefined;
+}
+
+interface props {}
```
Comment (ide) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]:
```suggestion
interface Props {}
```

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import * as Application from 'expo-application';
+import HeadingText from '../components/HeadingText';
+import MonoText from '../components/MonoText';
+
+interface ApplicationConstantsName {
+  name?: string;
+  value?: any;
+}
+
+interface mState {
```
Comment (ide) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]:
```suggestion
interface State {
```

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        <ApplicationConstants
+          name="Application applicationName"
+          value={Application.applicationName}></ApplicationConstants>
+        <ApplicationConstants
+          name="Application applicationId"
+          value={Application.applicationId}></ApplicationConstants>
+        <ApplicationConstants
+          name="Application androidId"
+          value={Application.androidId}></ApplicationConstants>
+        <View style={{ padding: 10 }}>
```
Comment (sjchmiela) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]: how about extracting this View construct to a component too? 🙂

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+interface ApplicationConstantsName {
+  name?: string;
+  value?: any;
+}
+
+interface State {
+  installReferrer?: string;
+  idForVendor?: string;
+  firstInstallDate?: string | undefined;
```
Comment (sjchmiela) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]: Doesn't `?` include `undefined` value as a valid value?

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,136 @@
+import React from 'react';
+import { ScrollView, View, Text, Button, Platform } from 'react-native';
+import * as Application from 'expo-application';
+import HeadingText from '../components/HeadingText';
+import MonoText from '../components/MonoText';
+
+interface ApplicationConstantsName {
```
Comment (sjchmiela) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]:
```suggestion
interface ApplicationConstant {
```

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+class ApplicationConstants extends React.Component<ApplicationConstantsName> {
+  render() {
+    let { name, value } = this.props;
+    if (typeof value === 'boolean') {
+      return (
+        <View style={{ marginBottom: 10 }}>
+          <HeadingText>{name}</HeadingText>
+          <MonoText> {String(value)}</MonoText>
```
Comment (sjchmiela) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]: Prettify file

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    }
+  }
+}
+
+export default class ApplicationScreen extends React.Component<{}, State> {
+  static navigationOptions = {
+    title: 'Application',
+  };
+
+  state: State = {
```
Comment (sjchmiela) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]: React should infer that it's of type `State` if you pass proper generic type to `React.Component<Props, State>` 🙂

**Review by ide — APPROVED** [review]

File: apps/native-component-list/src/screens/ApplicationScreen.tsx [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  componentDidMount() {
+    this._getInstallReferrer();
+    this._getIDFV();
+    this._getLastUpdateDate();
+    this._getFirstInstallDate();
+  }
+
+  _getInstallReferrer = () => {
+    Application.getInstallReferrerAsync().then(installReferrer => {
+      this.setState({ installReferrer: installReferrer });
```
Comment (sjchmiela) [inline apps/native-component-list/src/screens/ApplicationScreen.tsx]: property shorthand should work as well 🙂

**Pull request merged**